core/iwasm/libraries/libc-wasi/libc_wasi_wrapper.c

```diff
@@ -1021,24 +1024,20 @@ execute_interruptible_poll_oneoff(wasm_module_inst_t module_inst,
     bh_memcpy_s(in_copy, size_to_copy, in, size_to_copy);
 
     while (timeout == (__wasi_timestamp_t)-1 || elapsed <= timeout) {
-        elapsed += time_quant;
-
         /* update timeout for clock subscription events */
         update_clock_subscription_data(in_copy, nsubscriptions,
                                        min(time_quant, timeout - elapsed));
         err = wasmtime_ssp_poll_oneoff(curfds, in_copy, out, nsubscriptions,
                                        nevents);
+        elapsed += time_quant;
+
         if (err) {
             wasm_runtime_free(in_copy);
             return err;
         }
 
-        if (wasm_runtime_get_exception(module_inst) || *nevents > 0) {
+        if (get_suspend_flags(exec_env) & 0x01) {
```

Comment:
Should also handle `*nevents > 0`:
```
if (thread is terminated) {
    ...
}
else if (*nevents > 0) {
    wasm_runtime_free(in_copy);
    return __WASI_ESUCCESS;
}

Reply:
> I removed it on purpose since *nevents evaluates to 1 already at the first iteration and that ends the loop immediately

If that is the case, that means there was an event and we should indeed end the loop. That's how `poll()` works in general - it only blocks until either it reaches the timeout OR it detects an event.

Reply:
It seems like it's always set to `1`
https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2615

Reply:
> It seems like it's always set to `1`
> 
> https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2615

It is set to `1` only when `if (nsubscriptions == 1 && in[0].u.type == __WASI_EVENTTYPE_CLOCK) {`:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#LL2542C5-L2542C73

If `nsubscriptions > 1`, it will run into L2624:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2624

and `*nevents` may be increased
https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2738

https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2752

Reply:
so the condition should be something like:
```c
else if (*nevents > 0 && (nsubscriptions > 1 || in[0].u.type != __WASI_EVENTTYPE_CLOCK)) {
    wasm_runtime_free(in_copy);
    return __WASI_ESUCCESS;
}
```

Reply:
Had better `else if (*nevents > 0 && !(nsubscriptions == 1 && in[0].u.type == __WASI_EVENTTYPE_CLOCK))`

Reply:
> It seems like it's always set to `1`
> 
> https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2615

Seems that it is a bug of posix.c when `if (nsubscriptions == 1 && in[0].u.type == __WASI_EVENTTYPE_CLOCK) {`, it always returns 0 even when error occurs, which is different from the handling of the non `if (nsubscriptions == 1 && in[0].u.type == __WASI_EVENTTYPE_CLOCK) {`,  the latter sets `error` to 0 first, and converts `errno` to `error` when poll returns -1, and eventually returns `error`:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2712-L2715

My personal understanding is that we should also apply similar changes in the former, for example, change
https://github.com/bytecodealliance/wasm-micro-runtime/blob/d2772c4153a5ca27b33df3e3d5f0aa4c804234a1/core/iwasm/libraries/libc-wasi/sandboxed-system-primitives/src/posix.c#L2615-L2616
to:
```C
    *nevents = 1;
    if (out[0].error != 0)
        return convert_error(out[0].error);
    return 0;
```

And here, restore to
```C
if (thread is terminated) {
    ...
}
else if (*nevents > 0) {
    wasm_runtime_free(in_copy);
    return __WASI_ESUCCESS;
}
```
Not sure it is right, could you help review and test?

Reply:
It makes sense, but how does that make the `!(nsubscriptions == 1 && in[0].u.type == __WASI_EVENTTYPE_CLOCK)` part in the else unnecessary? With your proposed changes we modify the return code, but `*nevents` is still set to `1`, making the loop always return `__WASI_ESUCCESS` at the first iteration

Reply:
Seems you are right, we divide total timeout into many pieces, for the case of `nsubscriptions == 1 && in[0].u.type == __WASI_EVENTTYPE_CLOCK`, it should wait many times until the total timeout is reached.
Seems we should change the first part of `wasmtime_ssp_poll_oneoff` to:
```C
    *nevents = 1;
    if (out[0].error != 0)
        return convert_error(out[0].error);
    return 0;
```
and here keep:
```C
if (thread is terminated) {
    ...
}
else if (*nevents > 0 && !(nsubscriptions == 1 && in[0].u.type == __WASI_EVENTTYPE_CLOCK)) {
    wasm_runtime_free(in_copy);
    return __WASI_ESUCCESS;
}
```
Is that right?

Reply:
like this?
```c
else if (*nevents > 0) {
    all_outs_are_type_clock = true;
    for (i = 0; i < nsubscriptions; i++) {
        if (out[i].type != __WASI_EVENTTYPE_CLOCK)
            all_outs_are_type_clock = false;
    }

    if (!all_outs_are_type_clock) {
        wasm_runtime_free(in_copy);
        return __WASI_ESUCCESS;
    }
}
```

Reply:
Ok, I added that